lib/src/block_syntaxes/list_syntax.dart

```diff
@@ -89,16 +99,41 @@ abstract class ListSyntax extends BlockSyntax {
       } else if (tryMatch(hrPattern)) {
         // Horizontal rule takes precedence to a new list item.
         break;
-      } else if (tryMatch(ulPattern) || tryMatch(olPattern)) {
+      } else if ((isCheckboxListSubclass &&
+              (tryMatch(ulWithCheckBoxPattern) ||
+                  tryMatch(olWithCheckBoxPattern))) ||
+          tryMatch(ulPattern) ||
+          tryMatch(olPattern)) {
+        // The checkbox subclasse's patterns ([ulWithCheckBoxPattern] and
+        // [olWithCheckBoxPattern]) have 2 extra capturing groups at the 5
+        // position to capture the checkbox. This shift the other groups
+        // forward by 2 slots.
+        final cbGroupOffset = isCheckboxListSubclass ? 2 : 0;
         final precedingWhitespace = match![1]!;
```

Comment:
Oof, these `match!` are killing me. Since you're adding more in this PR, let's go ahead and make a local variable.

Our team has been usnig an idiom that I rather like: create a local variable that escapes the non-promotability of the other variable. Something like `final successfulMatch = match;` inside this if-block. And use that new variable below, so we don't have to keep null-checking it. Or better yet, we could rename the variable above (`late Match? match;`) to `possibleMatch`? What do you think?

Reply:
One more comment here.